### ISSUE TITLE
Response Time Percentiles

### ISSUE THREAD

**Issue by msmsimondean** [openzipkin/zipkin#2469]
**Feature:**
Aggregate spans to produce and display percentiles (e.g. p50, p90, p99, p99.9 etc.) for execution times of spans.  

**Rational**
So it is possible to view a summary of execution times rather than only being able to find and view individual spans.  

**Example Scenario**
The feature could work as follows:
1. Use Zipkin UI to search for spans based on name and/or tags
1. The UI shows percentiles for the 10 spans
1. Change the search limit from 10 spans to 1,000 spans
1. The UI shows percentiles for the 1,000 spans

**Prior Art**
* Gatling includes percentiles in it's perf test reports: https://gatling.io/docs/2.3/general/reports/

**Comment by codefromthecrypt** [comment]
aggregation of spans can be done in-process (like micrometer) or at ingest (like haystack metrics)

does this link help? if not why https://github.com/ExpediaDotCom/haystack-docker#to-start-zipkin-tracing-with-haystacks-trends-and-service-graph

**Comment by msmsimondean** [comment]
Thanks for the feedback @adriancole.  I've taken the in-process approach before.  One of the downsides is it's really difficult to aggregate across multiple instances of the same service.  I've ended up implementing a banding based approach as suggested on https://www.vividcortex.com/blog/why-percentiles-dont-work-the-way-you-think .  Downside is that the implementation ends up being specific to the web framework used (e.g. Spring Boot or Dropwizard) and so needs implementing multiple times if you use more than 1 framework in your estate.  

Haystack seems interesting, though it's quite a big addition just to compute percentiles.  

Are there any thoughts/plans to implement summary/aggregte level views in Zipkin in the future or will it always stay at the level of finding and then viewing individual traces?  Thanks!

**Comment by codefromthecrypt** [comment]
currently we have no other existing issue about aggregation on percentiles
besides this. to be honest with you is to say that this issue would need a
lot more buy in as it impacts all storage options.

right now we do have duration query but you would need to use a different
tool to decide which duration matters. that is the most honest assessment
of status quo.

if there is something about haystack that could be made easier I would ask
to give them that feedback. they are nice and earnest people. if matters
less if you are able to use their stuff..  rather a nice gesture to tell
them what's up.

**Comment by codefromthecrypt** [comment]
another possibility is using @jeqo's kafka storage.. you could do latency aggregation in the pipeline, but it might need some serious UI work to integrate the data. For example, the aggregation job could keep data local and then when querying for traces integrate the place in latency distribution info as a pseudo tag on readback. searching by distribution wouldn't work, but you could possibly see where it was anyway.

https://github.com/jeqo/zipkin-storage-kafka

**Comment by jkschneider** [comment]
> One of the downsides is it's really difficult to aggregate across multiple instances of the same service. 

This is true if you compute percentiles in the application before shipping them. Clearly percentile data itself is non-aggregable across dimensions. 

This is what percentile approximations off of discretized histogram data is designed to solve. See things like [histogram_quantile](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile). The individual histograms for several dimensions (say a host dimension) can be aggregated by pairwise summing the buckets and then performing the calculation off of the summed histogram.

Even percentiles may not be as useful as you folks think, depending on the application. If what you are really after as a business is saying "I want no more than 1% of my users to experience a greater than 10ms latency", once way of viewing this is that the 99th percentile needs to be less than 10ms. Another way of looking at this is to simply count how many requests are less than 10ms, and then  `(total_throughput - count_less_than_10ms) / total_throughput <= 0.01`. This latter approach is far cheaper than even computing percentile approximations over discretized histogram data.

Bottom line: percentile data needs to be computed off of either 100% sampled tracing data (haystack) or metrics telemetry (aggregated). Trying to reason about a distribution from downsampled data is a fool's errand.

**Comment by codefromthecrypt** [comment]
I don't think a change around this would occur in this repository
